ADApp/pluginTests/test_NDPluginOverlay.cpp

```diff
@@ -0,0 +1,255 @@
+    // Test an case with very large width
+    overlayTempCaseStr test7 = {0, 500, 500, 50, 50, 5000, 5000, 0, 255, 0, NDOverlayCross, NDOverlaySet, 2, {1024, 1024}, NDColorModeMono};
+    appendTestCase(&overlayTestCaseStrs, &test7); 
+    // Test an case with zero width
+    overlayTempCaseStr test8 = {0, 500, 500, 50, 50, 0, 0, 0, 255, 0, NDOverlayCross, NDOverlaySet, 2, {1024, 1024}, NDColorModeMono};
+    appendTestCase(&overlayTestCaseStrs, &test8); 
+}
+
+  ~OverlayPluginTestFixture()
+  {
+    delete arrayPool;
+    client.reset();
+    Overlay.reset();
+    driver.reset();
+    //delete downstream_plugin; // TODO: We can't delete a TestingPlugin because it tries to delete an asyn port which doesnt work
+  }
+
+};
+
+BOOST_FIXTURE_TEST_SUITE(OverlayPluginTests, OverlayPluginTestFixture)
+
+
+BOOST_AUTO_TEST_CASE(validate_input_NDArrays)
+{
+  BOOST_MESSAGE("Validating the input NDArrays " <<
+                "- if any of these fail the test-cases are also likely to fail");
+
+  overlayTestCaseStr *pStr = &overlayTestCaseStrs[0];
+  for (size_t i=0; i<overlayTestCaseStrs.size(); i++, pStr++)  {
+    for (size_t dim=0; dim<pStr->arrayDims.size(); dim++) {
+       BOOST_REQUIRE_EQUAL(pStr->arrayDims[dim], pStr->pArrays[0]->dims[dim].size);
+    }
+    BOOST_REQUIRE_EQUAL(pStr->pArrays[0]->dataType, NDFloat32);
+    BOOST_REQUIRE_EQUAL(pStr->arrayDims.size(), pStr->pArrays[0]->ndims);
+  }
+}
+
+
+BOOST_AUTO_TEST_CASE(basic_overlay_operation)
+{
+
+  overlayTestCaseStr *pStr = &overlayTestCaseStrs[0];
+  for (size_t i=0; i<overlayTestCaseStrs.size(); i++, pStr++)  {
+  
+    BOOST_MESSAGE("Test " << (i+1) << " rank: " << pStr->rank);
+
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayUseString,       1));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayPositionXString, pStr->positionX));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayPositionYString, pStr->positionY));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlaySizeXString,     pStr->sizeX));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlaySizeYString,     pStr->sizeY));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayWidthXString,    pStr->widthX));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayWidthYString,    pStr->widthY));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayShapeString,     pStr->shape));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayDrawModeString,  pStr->drawMode));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayRedString,       pStr->red));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayGreenString,     pStr->green));
+    BOOST_CHECK_NO_THROW(Overlay->write(NDPluginOverlayBlueString,      pStr->blue));
+
+    BOOST_CHECK_NO_THROW(Overlay->write(NDArrayCallbacksString, 1));
+    BOOST_CHECK_EQUAL(Overlay->readInt(NDArrayCallbacksString), 1);
+
+    Overlay->lock();
+    BOOST_CHECK_NO_THROW(Overlay->processCallbacks(pStr->pArrays[0]));
```

Comment:
Am I right to understand that the only real test here is whether the plugin processes the frame and the downstream plugin input counter increments?

I suppose generating test images and checking the output pixel-by-pixel would be a bit of an overhead... It might be made easier with an extension to the test utilities but that's another development!